xla/service/gpu/buffer_comparator.h

```diff
@@ -43,11 +44,31 @@ class BufferComparator {
   // See the implementation for the tolerance value.
   absl::StatusOr<bool> CompareEqual(se::Stream* stream,
                                     se::DeviceMemoryBase current,
-                                    se::DeviceMemoryBase expected) const;
+                                    se::DeviceMemoryBase expected);
+ private:
+  // Returns `true` if two buffers are equal, `false` otherwise.
+  template <typename ElementT>
+  absl::StatusOr<bool> DeviceCompare(std::string_view kernel_name,
+                                        void* kernel_symbol);
+
+  // Host side comparison code that does the same thing, but reports some of the
+  // differences as well. It only print logs for debugging.
+  //
+  // Returns true if no differences were seen, false otherwise.
+  template <typename ElementType, typename ComparisonType>
+  absl::StatusOr<bool> HostCompare();
+
+  template <typename ElementT, typename ComparisonT>
+  absl::StatusOr<bool> CompareEqualParameterized(
+    std::string_view kernel_name, void* kernel_symbol);
 
  private:
   Shape shape_;
-  HloModuleConfig config_;
+  float rtol_tolerance_;  // relative tolerance for comparison
+  bool verbose_;          // whether to print out error message on mismatch
+  se::Stream* stream_ = nullptr;
+  se::DeviceMemoryBase current_;
+  se::DeviceMemoryBase expected_;
```

Comment:
Could we not add these members please?
I think that Compare(a, b) is more readable then Compare() which reads a_ and b_.
In general in XLA multiple people prefer functional style now, because it shows dependencies instead of hiding them.
Also these are just used to do parameter passing in an implicit way. What if we wanted to make this class multi-threaded later? It wouldn't work in that case..

Reply:
OK, then I will rewind some of my refactoring changes: perhaps, I can group same parameters into some struct in order to avoid passing the same set of params over again (to make code a bit less verbose).